python/ray/tune/trial_executor.py

```diff
@@ -15,39 +16,75 @@
 logger = logging.getLogger(__name__)
 
 
+# Ideally we want to use @cache; but it's only available for python 3.9.
+# Caching is only helpful/correct for no autoscaler case.
 @lru_cache()
-def _get_warning_threshold() -> float:
+def _get_cluster_resources_no_autoscaler() -> Dict:
+    return ray.cluster_resources()
+
+
+def _can_fulfill_no_autoscaler(trial: Trial) -> bool:
+    """Calculates if there is enough resources for a PENDING trial.
+
+    For no autoscaler case.
+    """
+    assert trial.status == Trial.PENDING
+    total_cpu = trial.resources.cpu + trial.resources.extra_cpu
+    total_gpu = trial.resources.gpu + trial.resources.extra_gpu
+    return total_cpu <= _get_cluster_resources_no_autoscaler().get(
+        "CPU", 0) and total_gpu <= _get_cluster_resources_no_autoscaler().get(
+            "GPU", 0)
+
+
+@lru_cache()
+def _get_insufficient_resources_warning_threshold() -> float:
     if is_ray_cluster():
         return float(
             os.environ.get(
                 "TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S_AUTOSCALER", "60"))
     else:
+        # Set the default to 10s so that we don't prematurely determine that
+        # a cluster cannot fulfill the resources requirements.
         return float(
-            os.environ.get("TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S", "1"))
+            os.environ.get("TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S", "10"))
```

Comment:
It seems like on the product, this is being raised too frequently?

Reply:
The problem in current version of code is less about the frequency, but rather the assumption about condition checking is off. See this [line](https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/tune/trial_executor.py?L240). It incorrectly concludes if every time we arrive at this condition check with no running trial, we are not making progress - which is wrong.
E.g. [trial1(PENDING)] --> [trial1(TERMINATED), trial2(PENDING)] --> [trial1(TERMINATED), trial2(TERMINATED), trial3(PENDING)] etc. Every time, there is no running trial, but progress is made.

The current PR addresses this issue by changing the condition check to the length of all_trials, which corrects the bug.

As for the frequency, a warning message is printed for autoscaler case every 60s. For no autoscaler case, we will throw an error after 10s if the cluster doesn't have enough resources to fulfill a Trial, which is a hard fail(no repeated printing).

If 60s and 10s are too short, we can certainly elongate that. We just need to trade off with the time that user starts to wonder what is happening with pending trials without being informed.

Reply:
we should probably test this in the product.

Reply:
Sure. Although the nature of the bug is more within the core Tune code. See the response to Richard's comment above. I pasted a SxS comparison of w and w/o the change. Let me know if you want some more coverage. 

Also Richard, Ameer, if we need more time to settle this PR, should we first revert the master?

Reply:
Actually, I think this PR is good (with a small adjustment to the error message).